### PR TITLE
feat(dropdown-breadcrumbs): adds badge to dropdown, dropdown to breadcrumbs, dropdown to menu

### LINKED ISSUE #3491: Enhancement: menu breadcrumbs
@mcarrano, @doruskova and I discussed multi-level breadcrumbs, creating this issue to memorialize our convo as I'll be out 12/21 - 12/25. Here are the notes:

Screen Shot 2020-09-18 at 10 18 15 AM

* Breadcrumbs are not present at root
* When drilled in
  * Root menu's title is `All`
  * Each submenu gets a unique title that is NOT clickable
* Flip around the arrows for the back links ('< parent node' rather than 'parent node >')

To explore
• How to put tag inside dropdown

### PR DESCRIPTION
**Pull request by mattnolting**
closes #3491 #3657

This PR adds a `badge toggle` for `dropdowns`, `dropdown` w/`badge` in `breadcrumbs` and `badge` w/`dropdown` for `menu` navigation.

Badge toggle: https://patternfly-pr-3797.surge.sh/components/dropdown#badge-toggle
In dropdown: https://patternfly-pr-3797.surge.sh/components/breadcrumb#with-dropdown
In breadcrumbs: https://patternfly-pr-3797.surge.sh/components/menu#drilldown-with-breadcrumbs--level-1

### CODE REVIEW

**Review by doruskova — COMMENTED** [review]
It looks amazing! I don't have any comments. Thank you!

File: src/patternfly/components/Menu/examples/Menu.md [hunk 1]
```diff
@@ -316,6 +316,37 @@ import './Menu.css'
 {{> menu-drilldown-example menu-drilldown-example--menu__content--attribute='style="--pf-c-menu--m-drilldown--Width: 350px;"'}}
```
Comment (DaoDaoNoCode) [inline src/patternfly/components/Menu/examples/Menu.md:316]: Should this one and the 3 examples above also change to `menu--Drilldown`?

Comment (mattnolting) [reply src/patternfly/components/Menu/examples/Menu.md]: Nice catch! Updated the partial refs.

File: src/patternfly/components/Dropdown/dropdown-toggle--Badge.hbs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{{#> badge badge--modifier="pf-m-read"}}
+  {{#if dropdown-toggle--Badge--text}}
+    <span class="pf-c-dropdown__toggle-badge-text">{{dropdown-toggle--Badge--text}}</span>
+  {{else}}
+    <span class="pf-c-dropdown__toggle-badge-text">5</span>
+  {{/if}}
+  {{#unless dropdown-toggle--Badge--HasNoToggleIcon}}
```
Comment (mcoker) [inline src/patternfly/components/Dropdown/dropdown-toggle--Badge.hbs:7]: could this be rewritten as

```
<span class="pf-c-dropdown__toggle-badge-text">
  {{#if dropdown-toggle--Badge--text}}
    {{dropdown-toggle--Badge--text}}
  {{else}}
     5
  {{/if}}
</span>
```

File: src/patternfly/components/Dropdown/dropdown.scss [hunk 3]
```diff
@@ -114,6 +114,16 @@ $pf-c-dropdown--breakpoint-map: build-breakpoint-map("base", "sm", "md", "lg", "
   --pf-c-dropdown--c-divider--MarginTop: var(--pf-global--spacer--sm);
   --pf-c-dropdown--c-divider--MarginBottom: var(--pf-global--spacer--sm);
 
+  // badge toggle icon
+  --pf-c-dropdown__toggle--c-badge__toggle-icon--MarginLeft: var(--pf-global--spacer--xs);
+  --pf-c-dropdown__toggle-c--badge__toggle-icon--MarginRight: 0;
```
Comment (mcoker) [inline src/patternfly/components/Dropdown/dropdown.scss:119]:
```suggestion
  --pf-c-dropdown__toggle--c-badge__toggle-icon--MarginRight: 0;
```

File: src/patternfly/components/Dropdown/dropdown.scss [hunk 4]
```diff
@@ -318,6 +334,11 @@ $pf-c-dropdown--breakpoint-map: build-breakpoint-map("base", "sm", "md", "lg", "
     }
   }
 
+  > .pf-c-badge {
+    --pf-c-dropdown__toggle-icon--MarginLeft: var(--pf-c-dropdown__toggle--c-badge__toggle-icon--MarginLeft);
+    --pf-c-dropdown__toggle-icon--MarginRight: var(--pf-c-dropdown__toggle-c--badge__toggle-icon--MarginRight);
```
Comment (mcoker) [inline src/patternfly/components/Dropdown/dropdown.scss:339]:
```suggestion
    --pf-c-dropdown__toggle-icon--MarginRight: var(--pf-c-dropdown__toggle--c-badge__toggle-icon--MarginRight);
```

File: src/patternfly/components/Menu/menu.scss [hunk 5]
```diff
@@ -187,6 +199,28 @@
...
+  padding-right: var(--pf-c-menu__breadcrumb--PaddingRight);
+  padding-left: var(--pf-c-menu__breadcrumb--PaddingLeft);
+
+  .pf-c-breadcrumb__item {
+    font-size: var(--pf-c-menu__breadcrumb--c-breadcrumb__item--FontSize);
+  }
+
+  .pf-c-breadcrumb__heading {
+    --pf-c-breadcrumb__heading--FontSize: var(--pf-c-menu__breadcrumb--c-breadcrumb__heading--FontSize);
+  }
```
Comment (mcoker) [inline src/patternfly/components/Menu/menu.scss:216]: could `__item` and `__heading` be rewritten as

```
.pf-c-breadcrumb {
  --pf-c-breadcrumb__item--FontSize: var(--pf-c-menu__breadcrumb--c-breadcrumb__item--FontSize);
  --pf-c-breadcrumb__heading--FontSize: var(--pf-c-menu__breadcrumb--c-breadcrumb__heading--FontSize);
}
```

File: src/patternfly/components/Dropdown/examples/Dropdown.md [hunk 6]
```diff
@@ -156,6 +158,8 @@ The dropdown menu can contain either links or buttons, depending on the expected
 | `.pf-c-dropdown__toggle-text` | `<span>` | Defines the dropdown toggle text. **Required when text is present, adds truncation**. |
 | `.pf-c-dropdown__toggle-check` | `<label>` | Defines a checkbox in the toggle area of a split button dropdown. |
 | `.pf-c-dropdown__toggle-button` | `<button>` | Defines the toggle button for a split button dropdown. |
+| `.pf-c-dropdown__toggle-badge` | `<span>` | Defines the dropdown toggle badge. |
+| `.pf-c-dropdown__toggle-badge-text` | `<span>` | Defines the dropdown badge text. **Required**. |
```
Comment (mcoker) [inline src/patternfly/components/Dropdown/examples/Dropdown.md:162]: Do you still need these lines for `pf-c-dropdown__toggle-badge` and `pf-c-dropdown__toggle-badge-text`?

File: src/patternfly/components/Menu/menu.scss [hunk 7]
```diff
@@ -187,6 +199,25 @@
   }
 }
 
+// breadcrumb
+.pf-c-menu__breadcrumb {
+  display: flex;
+  align-items: center;
+  min-height: var(--pf-c-menu__breadcrumb--MinHeight);
```
Comment (mcoker) [inline src/patternfly/components/Menu/menu.scss:206]: could you use negative margins (-1 * --spacer--form-control) on the `pf-c-breadcrumb__dropdown` element instead?

Comment (mcoker) [reply src/patternfly/components/Menu/menu.scss:206]: then if the dropdown appears in some other context, the breadcrumb height won't change

**Comment by DaoDaoNoCode** [comment]
Screen Shot 2021-01-19 at 3 09 32 PM
A small question, is there a way to make the space between the right angle icon and every item consistent? As is shown in the picture, the space between badge and icon is larger than the space between text and icon. Is it safe to set the left and right padding (or even all the paddings) of the badge button to `0`?

**Comment by mattnolting** [comment]
> Screen Shot 2021-01-19 at 3 09 32 PM
> 
> A small question, is there a way to make the space between the right angle icon and every item consistent? As is shown in the picture, the space between badge and icon is larger than the space between text and icon. Is it safe to set the left and right padding (or even all the paddings) of the badge button to `0`?

@DaoDaoNoCode yep, we could do that. I'd like to see what design says and, if they want to update it, we can create a followup issue for that.

**Review by mcoker — APPROVED** [review]
LPTM! Great job!!

**Comment by mcoker** [comment]
> A small question, is there a way to make the space between the right angle icon and every item consistent? As is shown in the picture, the space between badge and icon is larger than the space between text and icon. Is it safe to set the left and right padding (or even all the paddings) of the badge button to 0?

I think the padding on the dropdown toggle is a good thing, as it increases the target area for the toggle. Though we could use negative left/right margins on the toggle to account for the space added between the `>` divider and the breadcrumb item, so that there is no space between the `>`'s that come before and after the toggle. 

```
.pf-c-breadcrumb__dropdown {
    margin-top: var(--pf-c-breadcrumb__dropdown--MarginTop);
    margin-bottom: var(--pf-c-breadcrumb__dropdown--MarginBottom);
    margin-left: calc(-1 * var(--pf-c-breadcrumb__item-divider--MarginRight));
    margin-right: calc(-1 * var(--pf-c-breadcrumb__item-divider--MarginRight));
}
```

Screen Shot 2021-01-19 at 2 47 02 PM
Screen Shot 2021-01-19 at 2 46 56 PM

**Review by DaoDaoNoCode — APPROVED** [review]
@mattnolting Very nice job! Thanks.

**Comment by DaoDaoNoCode** [comment]
Do we need to set a height for https://patternfly-pr-3797.surge.sh/components/breadcrumb#with-dropdown? Or it will shelter the documentation text under it.

**Pull request merged**